RFC for rustdoc jump to definition (#1)
GuillaumeGomez/rfcs: text/000-rustdoc-jump-to-definition.md

With this in mind, the suggested UI would be like this:

 * Links generated on idents in source code will jump to the ident definition.
 * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
 * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).

Basically, the source code pages' UI doesn't change.

# Drawbacks
[drawbacks]: #drawbacks

**The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?

Unfortunately, on this part, the best answer I can provide is that we will need to put clear limits to what we don't want. I think we need to keep rustdoc focused on its target: documentation. We provide source code pages, so this feature allows to browse them more efficiently by using information we already have. Anything more advanced would be very likely out-of-scope.

**Another concern is**: "why should we implement this in rustdoc? Aren't there already existing tools doing it?"

> A good answer to this is actually the following scenario: you are on <docs.rs> and you're looking at a crate documentation. You want to see how something is implemented and click on the `[src]` link. At this point, you encounter an item used in this page but not defined in this page. Do you want to clone the crate locally to check it out or go to github/gitlab to check it there or do you prefer having the links directly available?

Reviewer: That's actually a good point... why not just make the line numbers jump to github and let it be handled there?

Author: That brings a lot of other constraints: we'd need to support all websites' URL to go to a file line (if it's even supported). And it doesn't even work with private repositories too.

Reviewer 2: The other reason GitHub / GitLab links don't solve the problem is that they don't currently have code navigation supported for Rust: https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github.

**Last concern is about the maintenance**.

Any feature has a maintenance cost, this one included. It was suggested to put this feature in its own crate. The only part that can be extracted is the span map builder. But it would very likely be more of a problem than a solution considering that rustdoc API doesn't allow it easily.

One more argument about this is that the feature actually requires not that much code. It is split in two parts:

 * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/render/span_map.rs (which is basically a visitor gathering `span`)
 * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/highlight.rs#L725-L753 (for generating links)

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> [Github Code navigation feature](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github): Relying on external (and private) tools doesn't seem like a good idea. If they decide to shut it down, we can't do anything about it. In addition to that, not all projects are on github, so it excludes them. It also requires an internet connection.

Reviewer: If you have it locally, you already have the source locally, maybe offer URLs that open the local file path in that case so you can use vscode or vim or whatever?

Author: Not necessarily, it can be a dependency. And also, that would be a context switch (browser -> IDE).

Reviewer: dependencies are also downloaded locally, otherwise you couldn't have documented the main crate.

yea, but one could argue that even just "view source" is a context switch to an IDE XD I mean... in theory we could ship a vscode-in-the-browser for the "view source" feature

> 

Reviewer: There's also the "run r-a in wasm in the browser" variant

Author: The integration into rustdoc is much more complicated than just having a visitor and generating links. Also for accessibility, it would be a lot less good.

Reviewer: Not saying it is a solution to go to, but it should be listed in the alternatives section

rust-analyzer: This is a public opensource project. However, the setup isn't super easy (but it's a minor issue) and when using it on the rust compiler source code, it's quite resource-heavy. Also, if you want to look at a crate source code on <docs.rs>, you'd need to clone it locally to be able to use rust-analyzer on it.

# Prior art
[prior-art]: #prior-art

> As mentionned above, it is actually quite common for documentation tools to have a source viewer. Implementations vary a bit between them:

Reviewer: Looks like jump to definition only has prior art in a single one of these

Author: I didn't finish this section actually. ^^'

A good example where this would be very helpful is for the [rust compiler source code](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877): even if a fully set up IDE or github, it's very hard to go around without previous knowledge or help from someone else. It's also very common to have undocumented items or to just want to see how something is implemented. With the "jump to definition", it already improved quite a lot the browsing experience by allowing to jump super quickly by simply clicking on a item. The missing part is now to go back from the source code into the item documentation to make it complete.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

When going to the source code pages (as you can see already in the [compiler documentation](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877)), you will be able to click on idents to go to their definition.

Once at their definition, if you click on the item name, you'll arrive on its documentation page (if it has one, otherwise nothing will happen).

> An interesting thing to note: primitive types always link to documentation. Also, to make it more obvious if a link points to source code or to documentation, they should have different visual markers (different background/underline colors for example).

Reviewer 2: The discussion of different visual markers is a bit out of place. In general, considerations of UI are scattered throughout the doc. I think it would be better for most of the doc to focus on the actions people should be able to perform (go to definition; go to documentation), and have a separate section discussion UI possibilities to achieve that.

Author: Should I create a new section or should I simply remove it then?

Reviewer 2: I think you should create a new section to discuss UI.

Reviewer 3: Not sure if this has been mentioned already. I would advice against differentiating link types by color since that would harm accessibility for colorblind people (in case of red & blue: monochromacy / achromatopsia, blue cone monochromacy). Just something to keep in mind.

Author: We decided to not add new color for links (or background). Instead they will have an underline when hovered.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

> It doesn't require javascript since it's only generating links, making it work on both desktop and mobile, with or without javascript enabled.

Reviewer 2: ```suggestion
It doesn't require javascript since it's only generating links, making it work with or without javascript enabled.
```

The use or non-use of JavaScript doesn't affect the desktop/mobile distinction, since both desktop and mobile platforms support JavaScript.

So on the technical side, there is no big issues. You can see the first part of this RFC implemented [here](https://github.com/rust-lang/rust/pull/84176).

For the UI, we have the following constraints:

 * Must work on mobile and desktop: we cannot rely on mouse hover events.

>  * It's better if it works without javascript (it's important for accessibility, but also for maintenance reasons too: less JS code, less maintenance).

Reviewer 2: > less JS code, less maintenance

I disagree with this statement. It's entirely possible that it turns out generating links on the server side is more maintenance because it is tightly coupled to compiler internals, which frequently change.

The flip side is that the Rust community has more people who know Rust than those who know JavaScript, so putting more of the implementation in Rust means there are more people available to do maintenance.

Author: I disagree with your disagreement. :smile: 

Maybe I have a wrong idea but let me share what I had in mind while I wrote this statement: to have this information in the front-end, the back-end has to generate it in any case (either as a link or any other format). So deciding to use JS to handle this would mean that we still have code in the back-end but we would also have code in the front-end to treat it.

If that makes sense?

Reviewer 2: Yep, that makes sense - can you put it in the doc? :-D

With this in mind, the suggested UI would be like this:

 * Links generated on idents in source code will jump to the ident definition.
 * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
 * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).

Basically, the source code pages' UI doesn't change.

# Drawbacks
[drawbacks]: #drawbacks

> **The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?

Reviewer 2: This should also mention the most likely second feature: "find references." That is something I'd really like, and I think many other people would like too. We even got a request related to that [in Zulip yesterday](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/indirect.20type.20mentions).

Author: I don't consider it as a drawback but more like a potential extension.

Reviewer 2: I agree - I don't consider it a drawback. But given that (a) some of the team are worried about scope expansion, and (b) it is a scope expansion, you should address it somewhere in the doc, for instance to say "in the future, we are likely to implement 'find references' as well. Here's why that will fit well with the jump to definition feature we are implementing now"

A good answer to this is actually the following scenario: you are on <docs.rs> and you're looking at a crate documentation. You want to see how something is implemented and click on the `[src]` link. At this point, you encounter an item used in this page but not defined in this page. Do you want to clone the crate locally to check it out or go to github/gitlab to check it there or do you prefer having the links directly available?

**Last concern is about the maintenance**.

Any feature has a maintenance cost, this one included. It was suggested to put this feature in its own crate. The only part that can be extracted is the span map builder. But it would very likely be more of a problem than a solution considering that rustdoc API doesn't allow it easily.

One more argument about this is that the feature actually requires not that much code. It is split in two parts:

 * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/render/span_map.rs (which is basically a visitor gathering `span`)
 * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/highlight.rs#L725-L753 (for generating links)

> # Rationale and alternatives

Reviewer 2: Can you also discuss Google Code Search? They have implemented code navigation for Rust. For instance, see https://cs.opensource.google/fuchsia/fuchsia/+/main:src/factory/factory_store_providers/src/config.rs;l=62.

[rationale-and-alternatives]: #rationale-and-alternatives

[Github Code navigation feature](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github): Relying on external (and private) tools doesn't seem like a good idea. If they decide to shut it down, we can't do anything about it. In addition to that, a lot of projects are not on github, so it would exclude them. It also requires an internet connection.

rust-analyzer: This is a public opensource project. However, the setup isn't super easy (but it's a minor issue) and when using it on the rust compiler source code, it's quite resource-heavy. Also, if you want to look at a crate source code on <docs.rs>, you'd need to clone it locally to be able to use rust-analyzer on it.

> rust-analyzer in the browser: Another possibility would be to run it with wasm. However, it brings a few issues: wasm support in browsers is quite recent, it requires JS to work (obviously, but it's not a big problem), it would be quite heavy to run and it would complexify rustdoc build quite a lot.

Reviewer 2: > wasm support in browsers is quite recent

This looks like wasm is quite well supported: https://caniuse.com/wasm

Author: Good point. I'll remove this part.

| [doxygen](https://eigen.tuxfamily.org/dox/Matrix_8h_source.html) | yes | no | yes |
| [Elixir](https://elixir.bootlin.com) (used for linux kernel) | yes | yes | no |
| [Go](https://cs.opensource.google/go/go/+/go1.16.5:src/database/sql/sql.go;l=44) | yes | yes | no |
| [hexdocs](https://preview.hex.pm/preview/k6/0.0.1/show/lib/k6/template/web_socket.ex) | yes | no | no |
| python (sphynx/pydoc) | no (links to github) | no | no |
| [ruby](https://ruby-doc.org/stdlib-3.1.2/libdoc/delegate/rdoc/Delegator.html#method-i-__raise__) | yes (directly on the page) | yes/no (it is sometimes in examples) | no |
| [perl](https://metacpan.org/dist/DBD-SQLite/source/lib/DBD/SQLite.pm) | yes | yes | yes |
| [haskell](https://devdocs.io/haskell~9/libraries/binary-0.8.9.0/data-binary-builder#v:append) | yes | yes | no |
| [opal](https://opalrb.com/docs/api/v1.4.1/stdlib/Buffer) | yes (directly on the page) | no | yes |

# Unresolved questions
[unresolved-questions]: #unresolved-questions

Is it the best way to make the links to go to an item documentation like this? Maybe something adding a small icon to click right besides it should be better?

> Should we add links to documentation to all items (impl blocks, methods, trait/impl associated items) or not?

Reviewer 2: I think all items in source code that have documentation should have a way to go from the source view to the documentation.

Author: I think so too. I'll move it.

Rustdoc: jump to definition

- Feature Name: `jump_to_definition`
- Start Date: 2021-12-09
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#89095](https://github.com/rust-lang/rust/issues/89095)

# Summary
[summary]: #summary

Generate links on idents in the rustdoc source code pages which links to the item's definition and documentation (if available).

> # Motivation

Reviewer 4: issue: the contents of this section are much much more than just "motivation", i think this rfc needs to be organized a bit more

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

When going to the source code pages (as you can see already in the [compiler documentation](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877)), you will be able to click on idents to go to their definition.

Once at their definition, if you click on the item name, you will arrive on its documentation page (if it has one, otherwise nothing will happen).

All items with documentation should have a link to their documentation page (like impl blocks, methods, trait/impl associated items, etc).

An interesting thing to note: primitive types always link to documentation.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

It doesn't require javascript since it's only generating links, making it work with or without javascript enabled.

> So on the technical side, there is no big issues. You can see the first part of this RFC implemented [here](https://github.com/rust-lang/rust/pull/84176).

Reviewer 4: issue: this doesn't really belong here, also language like "there are no big issues" is presupposing what the issues _are_

Author: Based on what is described in this RFC. I'll add this sentence to make it more clear.

 * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).

Another thing which would be nice: to make it more obvious if a link points to source code or to documentation, they should have different visual markers (different background/underline colors for example).

Basically, the source code pages' UI doesn't change.

As for the size difference for the generated documentation, here are a few examples:

| crate name | without the feature (in bytes) | with the feature (in bytes) | diff |
| ---------- | ------------------------------ | --------------------------- | ---- |
| std | 11.788.288 | 13.172.736 | 11.7% |
| tract_core | 4.276.224 | 5.554.176 | 29.9% |
| stm32f4 | 22.335.488 | 26.984.448 | 20.8% |
| image | 4.927.488 | 6.045.696 | 22.7% |

# Drawbacks
[drawbacks]: #drawbacks

**The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?

> Unfortunately, on this part, the best answer I can provide is that we will need to put clear limits to what we don't want. I think we need to keep rustdoc focused on its target: documentation. We provide source code pages, so this feature allows to browse them more efficiently by using information we already have. Anything more advanced would be very likely out-of-scope.

Reviewer 4: _we need to do this in this RFC_, we cannot just say we will do this in this RFC. Strictly defining scope should go in the reference level section

Author: Actually about this I have a different approach. This RFC defines what it should implement for itself:

 * Jump to def
 * Jump back to doc
 * How the links UI should look like
 * How the links should be generated
 
Trying to define any further change seems to be out of scope no? Or did I misunderstood what you said?

Reviewer 4: No, you should explicitly determine what is not in scope from this RFC. The list of four points you have provided can actually be rather broad.

We also need to define: jump to def for _which_ nodes. Types? Expressions to their types? Everything? We _really really_ need to be more specific here.

Author: I wrote in the "Reference-level explanation" section:

> All items with documentation should have a link to their documentation page (like impl blocks, methods, trait/impl associated items, etc).

Is this enough (and the right place) or do you see a case that we don't want that would match this definition?

**Another concern is**: "why should we implement this in rustdoc? Aren't there already existing tools doing it?"

A good answer to this is actually the following scenario: you are on <docs.rs> and you're looking at a crate documentation. You want to see how something is implemented and click on the `[src]` link. At this point, you encounter an item used in this page but not defined in this page. Do you want to clone the crate locally to check it out or go to github/gitlab to check it there or do you prefer having the links directly available?

> 

Reviewer 4: another potential concern is the size of the generated code, it's worth addressing

Author: Good point!

Author: Oh actually it's already mentioned in `UI`. I really need to clean this up...

rust-analyzer: This is a public opensource project. However, the setup isn't super easy (but it's a minor issue) and when using it on the rust compiler source code, it's quite resource-heavy. Also, if you want to look at a crate source code on <docs.rs>, you'd need to clone it locally to be able to use rust-analyzer on it.

rust-analyzer in the browser: Another possibility would be to run it with wasm. However, it brings a few issues: it requires JS to work (obviously, but it's not a big problem), it would be quite heavy to run and it would complexify rustdoc build quite a lot.

External library (like `cpan` or `tree-sitter`): They would work to generate links to another source code location. However, they bring their downsides as well: they would need to be kept up-to-date with rust syntax evolution (if any) and making links back to documentation would be very tricky because some information is only internal to rustdoc (like how to differentiate between different `impl blocks` or different blanket trait implementations like `impl Trait<T: Clone> for T` and `impl Trait<T: Copy> for T`).

> # Prior art

Reviewer 4: uh, https://github.com/rust-dev-tools/cargo-src

Author: I didn't mention it because it uses rls which is deprecated and hasn't been updated in years.

Reviewer 4: It's still prior art!

Author: And done.

Rustdoc: jump to definition

- Feature Name: `jump_to_definition`
- Start Date: 2021-12-09
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#89095](https://github.com/rust-lang/rust/issues/89095)

# Summary
[summary]: #summary

Generate links on idents in the rustdoc source code pages which links to the item's definition and documentation (if available).

# Motivation
[motivation]: #motivation

This RFC proposes to generate links in the rustdoc source code pages (the ones you end up on when clicking on the `source` links) to allow you to jump to an item definition or back to its documentation page.

> The goal of this RFC is to greatly improve the browing experience of the source page (especially when looking for private items).

Reviewer 4: This section is still not a motivation.

The motivation is one of the most important parts of the RFC. It should not be written presupposing a conclusion. It is not a place to simply describe the feature. It is a place to, from the ground up, explain _why_ such a feature is desired.

Reviewer 4: It should also be careful explaining who this is for. Is this primarily for users of a crate looking to understand a type better? Or is it primarily for maintainers of the crate? That affects the design of this feature greatly.

For example, if it is primarily for users of a crate; to me it feels like we ought to focus on links in code examples _first_ (but that has its own complicated design problems). If it's primarily for maintainers, that's the scope expansion I'm worried about, rustdoc has historically been a tool for crate _users_ and if we're switching that around we need to be super super explicit about that.

Reviewer 4: cc @jsha who may have some thoughts about picking a target user base

Author: > This section is still not a motivation.
>
> The motivation is one of the most important parts of the RFC. It should not be written presupposing a conclusion. It is not a place to simply describe the feature. It is a place to, from the ground up, explain why such a feature is desired.

Ok, I think I understand my misunderstanding. The motivation should be something like "The source code pages usefulness is limited because it doesn't allow to actually browse efficiently through it". Or something along the line. Dissertation is not my forte so I'll need some time to come up with something more readable but that seems like a better start.

> It should also be careful explaining who this is for. Is this primarily for users of a crate looking to understand a type better? Or is it primarily for maintainers of the crate? That affects the design of this feature greatly.
> ...

Interestingly enough, the primary target I have for this feature is users, not crate maintainers, but we arrive at two different conclusions. As a user of a crate, I want to be able to understand how its internals work, and to do so I currently need to clone it locally because I can't easily "follow a lead" with the current source code pages. Which becomes quickly annoying when this same crate uses dependencies, etc.

Side note: I didn't think about adding links in examples. That could be a nice improvement too, but out of scope in the current case (still keeping it in mind for later though).

> Since a video is worth a thousand words:

Reviewer 4: Not for RFCs, RFCs should be readable. It's fine to link to a video but please please do not use a video to replace text contents, or write something that claims to do so. Let's just link to this in the guide-level explanation as a demo.

Author: The wording is indeed misleading. What I tried to express was: "you had the description, here you can see an example in action". Also, it should not be in the motivation section.

Once you found what you're looking for, having a link to an item's documentation from the source code allows to add the missing connection between the source code viewer and the doc pages.

Another important note is that if documentation is split in multiple places or requires an external tool for the source code navigation, it's actually slowing down our users quite a lot.

A good example where this would be very helpful is for the [rust compiler source code](https://doc.rust-lang.org/1.57.0/nightly-rustc/src/rustc_middle/hir/map/mod.rs.html#872-877): even if a fully set up IDE or github, it's very hard to go around without previous knowledge or help from someone else. It's also very common to have undocumented items or to just want to see how something is implemented. With the "jump to definition", it already improved quite a lot the browsing experience by allowing to jump super quickly by simply clicking on a item. The missing part is now to go back from the source code into the item documentation to make it complete.

> # Reference-level explanation

Reviewer 4: This section really needs some work.

At a minimum:

A list of exactly what kinds of links are happening. Type names to types, function names to functions, trait names to traits, macro names to macros, probably. const names to consts and static names to statics? Should generic types have each individual part of the generic type link correctly?

Are we also doing variable names to variable types or their local variable definitions? If we're linking them to their types (which is what it _seems_ like from this RFC), how about variables involving generics? And should we special case smart pointers (e.g. should a variable with type `&Foo` link to `Foo`? What about `Box<Foo>`? Where do we draw the line?)

And what about doc examples? It would seem that JTD in doc examples would be far more useful than in source, but also far trickier to implement especially since _currently_ `rustdoc` does not compile examples except when running tests.

This is kinda what I mean by pin down the scope, this RFC is _still_ rather broad and underspecified, there's a lot of work that needs to be done to pin down stuff like the above. The above is what I could think of quickly whilst writing this comment, we need to spend time brainstorming to figure out all potential edges of this figure.

We should have a subsection of this section that talks about "implementation details" too, containing the stuff you have below. But first, we ought to pin down the design.

[reference-level-explanation]: #reference-level-explanation

> It doesn't require javascript since it's only generating links, making it work with or without javascript enabled.

Reviewer 4: nit: this sentence reads really weirdly

I'd imagine something like the following:

> This feature can be implemented without JavaScript additions to Rustdoc as pure link generation; so it will work without JS enabled. Rustdoc already has an internal link generation system that can be used here.

The sentence about "all items with documentation" would be covered by stuff written before this because of my comment above. The sentence after that should go away.

For the link generation, we use the already existing system in rustdoc. Meaning their generation will follow the same rules as the link in the documentation pages.

All items with documentation should have a link to their documentation page (like impl blocks, methods, trait/impl associated items, etc).

> So on the technical side, there is no big issues (based on what is described in this RFC). A few parts of this RFC are implemented in the following pull requests:

Reviewer 4: This sentence is still completely out of place, an RFC does not get to pre-claim "there are no big issues". The point of the RFC process is to figure out there are no big issues.

Author: Ok, I'll remove it entirely then.

 * It's better if it works without javascript (it's important for accessibility, but also for maintenance reasons too: less JS code, less maintenance).

Some extra explanations about the statement "less JS code, less maintenance": To have this information in the front-end, the back-end has to generate it in any case (either as a link or any other format). So deciding to use JS to handle this would mean that we still have code in the back-end but we would also have code in the front-end to treat it.

With this in mind, the suggested UI would be like this:

 * Links generated on idents in source code will jump to the ident definition.
 * Links generated on items' definition will jump to their documentation page (if any, private items don't have one by default).
 * On mobile, these links will have dotted underline (because you don't have a cursor which change when hovering them).

Another thing which would be nice: to make it more obvious if a link points to source code or to documentation, they should have different visual markers (different background/underline colors for example).

Basically, the source code pages' UI doesn't change.

# Drawbacks
[drawbacks]: #drawbacks

> **The biggest concern is actually about "scope expansion"**. If we add this feature, are we doing too much? Rustdoc is supposed to be about documentation, so is providing more information into the source code page really necessary?

Reviewer 4: style nit: let's not turn the list of concerns into a big split-up sentence. Just have headings like "scope expansion:", "Other tools", etc.

> An answer about this is that it actually makes source code browsing much more pleasant and convenient. Multiple other documentation tools provide this feature for this reason so it doesn't seem like it would be out of scope in this regard.

Reviewer 4: This doesn't answer the question. "Rustdoc is about documentation, isn't this scope expansion?" is not answered by "this makes source code browsing more pleasant". In fact, it makes it sound _more_ like scope expansion since you answered a question about one field by mentioning another. It is not at all clear that rustdoc is or should be a tool for browsing the source. Just because it currently _contains_ source is insufficient, it can be a tool that contains source without being a tool for _browsing_ source.

This section (and the motivation section) needs to talk about how source code browsing is related to Rustdoc's job, and why source code browsing is useful for navigating docs.

The thing is, if the docs are good, you shouldn't _need_ to look at the source. This doesn't play out in practice that often but it's an easy argument to make against source code browsing being important. As a community we can (and somewhat have) set our norms towards having docs that don't require that. Would this cause a change to that norm? Would it make people want to write their source code so that people trying to _use_ the library can read it better (this can be a problem! writing a library so that it is readable to maintainers is different from writing one readable to users!).

| [Elixir](https://elixir.bootlin.com) (used for linux kernel) | yes | yes | no |
| [Go](https://cs.opensource.google/go/go/+/go1.16.5:src/database/sql/sql.go;l=44) | yes | yes | no |
| [hexdocs](https://preview.hex.pm/preview/k6/0.0.1/show/lib/k6/template/web_socket.ex) | yes | no | no |
| python (sphynx/pydoc) | no (links to github) | no | no |
| [ruby](https://ruby-doc.org/stdlib-3.1.2/libdoc/delegate/rdoc/Delegator.html#method-i-__raise__) | yes (directly on the page) | yes/no (it is sometimes in examples) | no |
| [perl](https://metacpan.org/dist/DBD-SQLite/source/lib/DBD/SQLite.pm) | yes | yes | yes |
| [haskell](https://devdocs.io/haskell~9/libraries/binary-0.8.9.0/data-binary-builder#v:append) | yes | yes | no |
| [opal](https://opalrb.com/docs/api/v1.4.1/stdlib/Buffer) | yes (directly on the page) | no | yes |
| [cargo-src](https://github.com/rust-dev-tools/cargo-src) | yes | yes | yes |

# Unresolved questions
[unresolved-questions]: #unresolved-questions

Is it the best way to make the links to go to an item documentation like this? Maybe something adding a small icon to click right besides it should be better?

# Potential extensions

> A potential extension would be "find references". It would allow users to check where the given item is being used. We already have all the information needed for this feature, so it's mostly how to use it and show it to the end users. However, if we decide to move forward with it, it will be part of another RFC.

Reviewer 4: this is exactly what i'm worried about wrt scope expansion, fwiw.

Rustdoc is not a source code browser tool. It has source in its capacity to _aid_ its purpose as a doc viewer. It is still insufficiently motivated (in this RFC) that better source code browsing is important for documentation viewing, because source is already very much a secondary purpose of rustdoc that it uses as an aid.

Going from there to reverse-search seems like an even bigger leap, and it only makes sense _if_ we have presupposed that Rustdoc is equally a tool for browsing source. And if we are deciding that, we need to decide that _very clearly_ in this RFC, and not have it just happen by accident.

The worry I have is that we are turning rustdoc from a tool designed for the _users_ of a crate to a tool designed for the _maintainers_ of a crate (and still the users, but now it's both). And that's a pretty big scope expansion.

Author: Just like prior-art, I added it because it was mentioned as a possible extension. Accepting this RFC doesn't necessarily mean that the feature will evolve any further. But if you prefer that I don't mention it, it's just as fine for me.

Here is a video showing the feature in action:

https://user-images.githubusercontent.com/3050060/114622354-21307b00-9cae-11eb-834d-f6d8178a37bd.mp4

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

This feature **only** targets the source code pages.

It can be implemented without JavaScript additions to Rustdoc as pure link generation; so it will work without JS enabled. Rustdoc already has an internal link generation system that can be used here.

## Link generation

There are two kinds of links:

 * Jump to documentation.
 * Jump to definition.

### Jump to documentation

> All items with a documentation page should have a link to it (like methods, trait/impl associated items, etc). However, fields, variants and impl blocks should not (explications below).

Author: Yes, it's much better, thanks!

The link is generated on their ident/name where they are declared, not when they are in use. For example:

```rust
pub struct Foo;

let x = Foo;
```

On `Foo` definition (`pub struct Foo`), we generate a link to its documentation page whereas on its usage (`x = Foo`), we generate a link to jump to its definition.

This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.

For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.

### Jump to definition

A small preamble before listing the items we should generate links for: if an item which should get a link is preceded by a path, the whole path is used as a link. For example:

```rust
let x: b::c::D; // `b::c::D` is the link to `D`.
foo::some_fn(); // `foo::some_fn` is the link to `some_fn`.
```

Anything not specifically mentioned should not get a link generated for itself.

#### Local variables

> Local variables should get a link to their definition (not their type):

Author: Oh I see. I originally fixed this by hiding the compiler errors because they don't impact the final output although errors are emitted. It was reverted afterward so they're displayed again. I'm too sure what to do about this.

> Local variables should get a link to their definition (not their type):

Author: Absolutely. I just wanted to know first if we wanted to propose a solution in the RFC or not. I plan to add this and the "up limit" to the RFC.

> Local variables should get a link to their definition (not their type):

Reviewer 2: When you're not sure what you should do about a problem, you should write that in the RFC. In particular, you should write:

 - what the problem is
 - what its impacts are (for instance, does this cause all crates to fail to build with this flag? only some crates?)
 - potentially some ideas about how to fix it, with the tradeoffs for those ideas

But if you don't have ideas for how to fix it, it's still important to write down the problem so that other people can help you figure out solutions (and so they don't have to discover the problem themselves from scratch).

Here is a video showing the feature in action:

https://user-images.githubusercontent.com/3050060/114622354-21307b00-9cae-11eb-834d-f6d8178a37bd.mp4

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

This feature **only** targets the source code pages.

It will be implemented without JavaScript additions to Rustdoc as pure link generation; so it will work without JS enabled. Rustdoc already has an internal link generation system that can be used here.

## Link generation

There are two kinds of links:

 * Jump to documentation.
 * Jump to definition.

### Jump to documentation

All items with a documentation page will have a link to it (like methods, trait/impl associated items, etc). However, fields, variants and impl blocks won't (explications below).

The link will be generated on their ident/name where they are declared, not when they are in use. For example:

```rust
pub struct Foo;

let x = Foo;
```

On `Foo` definition (`pub struct Foo`), we will generate a link to its documentation page whereas on its usage (`x = Foo`), we will generate a link to jump to its definition.

> This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we will generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.

Reviewer 3: Well, to be fair we could add a jump-to-documentation link on the `impl` keyword. Is that a good idea? If so, you could add it here to this paragraph or to the future extensions section, not sure which is best.

Author: I thought about it but it felt like it wouldn't be coherent. But that could work indeed! Let's see if others agree with you. If so I'll update.

Author: I added it into the unresolved questions section so it's not forgotten.

For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.

### Jump to definition

> A small preamble before listing the items we should generate links for: if an item which will get a link is preceded by a path, each part of the path will be a link to its target. For example:

Reviewer 3: ```suggestion
We will generate link for each path segment of a path to its respective definition. For example:
```

The phrasing of this paragraph confuses me a bit. Minor: I don't understand what “preamble” means in this context.
More significantly, however: You write ”*an item […] preceded by a path*” and – from what I infer – call the `D` in the example below an *item* and the `b::c` a *path*.
The quote is not correct, the whole snippet `b::c::D` is a path and `b`, `c` & `D` are all path *segments*. An item cannot be “*preceded by a path*”. `D` is not an item, it is a path segment referring to an item.
I don't want to sound overly pedantic.

Obviously my new phrasing overlaps a bit with the subsection **Jump to definition**/**Paths**. So one or the other can be cut down in size, I'd say.

Author: Preventing confusion is a very good thing, good catch!

```rust
fn f(mut a: String) {
    a.push_str("something"); // `a` get a link just like `push_str`!
}
```

The reasons about supporting local variables are as follow:

 * We want to linkify global variables, because they may be far away or in another file.
 * If you're looking at a variable reference in a function, you don't know whether it's a local variable or a global one (or one from an enclosing scope). By linkifying variables we make it easy for people to find out the answer. If we linkify only some variables (the global ones), it will appear weird that some are linkified and some are not. So for coherency, it will be done for all local variables.

#### Primitive types

No primitive type will get a link.

#### Fields

They will not get a link either.

#### Variants

They will get a link, but only on their usage (as already specified in the "jump to documentation" section):

```rust
enum Enum {
    Variant, // No link here.
}

// `Enum` links to `Enum` and `Variant` is the link to `Enum::Variant` definition.
let x = Enum::Variant;
```

#### Imports

Imports will have links as well, following the [Paths](#Paths) rules.

#### Types

> Whenever a type appears, we will link to its definition. For example:

Reviewer 3: ```suggestion
Whenever we encounter a type path that refer to types defined by an item (this excludes type parameters), we will link to its definition. For example:
```

Minor nit: To be technically more correct, you want to refer to *type paths* here since arbitrary *types* includes `dyn Thing`, `*const ()`, `[i32]` and the like.
The phrasing of *refer to types defined by an item* can probably be improved. As the example shows, we don't want to generate links for the type parameters `F` and `T`. So let's mention it in prose.

Author: :+1:

```rust
let x: String = "a".to_owned(); // we generate a link on `String`.
let y: Foo<Bar>; // We generate a link on both `Foo` and `Bar`.

fn foo<F: Display, T: F + Debug>(f: F, t: T) {} // We generate a link on `Display` and `Debug`.
```

#### Functions/methods

Whenever a function/method is present, we will generate a link to it. Example:

```rust

> let x: Fn() = some_fn; // We will link to `some_fn`.

Reviewer 3: ```suggestion
let x: fn() = some_fn; // We will link to `some_fn`.
```

Author: Outch, good catch!

A discussion about this suggested to instead add a new argument in the URL (`?source=true`) which would automatically click on the source link of the item on its documentation page. The problem is that private elements (unless the documentation is generated with `--document-private-items`) don't have a documentation page, which reduces the possibilities of this solution quite a lot. This fix would also only worked on web platform implementing it as it requires a server to pick a compatible version (in the semver meaning) for this crate. Another problem is: what to do if there is no compatible version (in the semver meaning still) available? For example, if there was only a `0.15` version published, `0.14.x` and `0.16.y` are not compatible to it. Then it would very likely just don't redirect and leave the user on the 404 page.

The idea here would be to follow the current URL strategy in use for cross-crate items: generate a URL relative to the root path. If `--extern-html-root-url` isn't used, then the root-path is the one provided to rustdoc where to build documentation.

> This is somewhat the same answer for `cargo --no-deps`: unless rustdoc actually checks if the dependencies' locations are empty or not, there is no way for rustdoc to know if they're present. As such, it should simply assume that they are present and generate links for them as it does for the foreign items in the documentation pages.

Reviewer 3: If I understand the whole *cross-crate links* section correctly, you seem to be saying that it is not possible at all for rustdoc to detect if a to-be-generated link would point to a nonexistent location.

However, from my limited experience I've gathered from writing several <kbd>A-cross-crate-reexport</kbd> PRs, I've learned that rustdoc *does* know if a cross-crate link does not resolve. Maybe that's not always the case but at least it is for code using [`html::format::href()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/html/format/fn.href.html).

So from what I know, rustdoc should be able to “gray out” / not generate (most? all?) links that would be dead. Specifically, the following paragraph should be updated to mention this:

> As such, it should simply assume that they are present and generate links for them as it does for the foreign items in the documentation pages.

Am I missing something major?

Author: I tried to mean not the cross-crate item itself but the actual HTML file destination. As I mentioned, you can run `cargo --no-deps` and in this case, only your crate will be documented and not its dependencies. As such, any link to any dependency will result in a 404. Does it clear the confusion? How could I rewrite it to make it less confusing?

Author: I tried to improve the wording and explanations of this section.

The link is generated on their ident/name where they are declared, not when they are in use. For example:

```rust
pub struct Foo;

let x = Foo;
```

On `Foo` definition (`pub struct Foo`), we generate a link to its documentation page whereas on its usage (`x = Foo`), we generate a link to jump to its definition.

This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.

For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.

### Jump to definition

A small preamble before listing the items we should generate links for: if an item which should get a link is preceded by a path, the whole path is used as a link. For example:

```rust
let x: b::c::D; // `b::c::D` is the link to `D`.
foo::some_fn(); // `foo::some_fn` is the link to `some_fn`.
```

Anything not specifically mentioned should not get a link generated for itself (such as variable name):

```rust
fn f(mut a: String) {

>     a.push_str("something"); // `a` should not get a link but `push_str` should!

Reviewer 4: We should probably have an explicit section about "local variables" saying that they do _not_ JTD

Author: That would make it more logical considering there are two sections for primitive types and fields.

Reviewer 2: Under the current implementation, local variables _do_ get a JTD link. For instance in https://rustdoc.crud.net/imperio/jump-to-def-regex/src/regex/pikevm.rs.html#205, at_next links to a `let` statement. Either the implementation is incorrect or the description here in the RFC that "`a` should not get a link" is incorrect.

Author: The implementation is incorrect. A few things are still "rough" and need some polishing. But we also all need to agree on what should or should not get links. Personally I think it's too much for variables but I'm curious to see if others share the opposite point of view.

Reviewer 2: I am in favor of supporting local variables. Here's why:

- We should definitely linkify global variables, because they may be far away or in another file.
- If you're looking at a variable reference in a function, you don't know whether it's a local variable or a global one (or one from an enclosing scope). By linkifying variables we make it easy for people to find out the answer. If we linkify only some variables (the global ones), we will leave people scratching their head about why some things are linkified and some are not.

Author: This is a fair argument. Should I put this argument directly in the section for local variables too? Having arguments alongside such cases could be useful but not sure if it's supposed to be in the RFC directly.

Reviewer 2: Yes, I think it makes sense to put the argument in the local variable section for now. It may make sense to reorganize it elsewhere based on subsequent edits.

#### Local variables

> Local variables should get a link:

Reviewer 4: Should be clear that you mean a link to their definition (not their type) here.

Author: Oh indeed! Thanks!

[SourceGraph](https://sourcegraph.com/github.com/rust-lang/rustc-demangle@2811a1ad6f7c8bead2ef3671e4fdc10de1553e96/-/blob/src/v0.rs?L65): SourceGraph uses [Language Server Index Format (LSIF)](https://github.com/Microsoft/language-server-protocol/blob/main/indexFormat/specification.md) to extract info from a language server and make it available in a static format. And rust-analyzer bindings for LSIF already exist. The problem is that it would require a live server to work, which is a blocker to be integrated into rustdoc as it must work serverless.

> Another argument in favour of not relying on an external tool is to have support for cross-crate links as it is a very important part of any rust project. For example, if you're browsing a crate "A" and you encounter an item coming from crate "B" (because crates often delegate significant portions of their implementations to sub-crates, for compile performance and other reasons), it seems obvious that it should be able to link to it. This means any tool that does a good job of navigating Rust source code needs to be aware of that code's dependencies, and how to link directly to source code pages within those dependencies. For instance, `askama_shared/filters/json.rs.html` needs to link to `serde::Serialize`. Rustdoc can do that because it builds a copy of all dependencies when building documentation, and it can use reliable mapping from (crate, version) to a URL pointing to the source code. GitHub or Google Code Search can't really do this mppaing while staying within their own code navigation system, because there is not a reliable mapping from (crate, version) to (repository URL, tag).

Reviewer 4: I don't think this is a particularly solid justification. It is important for rustdoc to know where dep docs are, and it already does! That does not explain why it needs to have it in its source view.

I might be misunderstanding this paragraph but it doesn't really seem to add new justification.
